OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+    // the GetInstanceProcAddr function of the next layer (or the driver if this is the last layer).
+    while (layer_create_info != nullptr &&
+           (layer_create_info->sType != VK_STRUCTURE_TYPE_LOADER_INSTANCE_CREATE_INFO ||
+            layer_create_info->function != VK_LAYER_LINK_INFO)) {
+      layer_create_info = absl::bit_cast<VkLayerInstanceCreateInfo*>(layer_create_info->pNext);
+    }
+
+    if (layer_create_info == nullptr) {
+      return VK_ERROR_INITIALIZATION_FAILED;
+    }
+
+    InitVulkanLayerProducerIfNecessary();
+
+    PFN_vkGetInstanceProcAddr next_get_instance_proc_addr_function =
+        layer_create_info->u.pLayerInfo->pfnNextGetInstanceProcAddr;
+
+    // Advance linkage for next layer.
+    layer_create_info->u.pLayerInfo = layer_create_info->u.pLayerInfo->pNext;
+
+    // Need to call vkCreateInstance down the chain to actually create the
+    // instance, as we need the instance to be alive in the create instance dispatch table.
+    auto create_instance = absl::bit_cast<PFN_vkCreateInstance>(
+        next_get_instance_proc_addr_function(VK_NULL_HANDLE, "vkCreateInstance"));
+    VkResult result = create_instance(create_info, allocator, instance);
+
+    dispatch_table_.CreateInstanceDispatchTable(*instance, next_get_instance_proc_addr_function);
+
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnCreateDevice(VkPhysicalDevice /*physical_device*/ physical_device,
+                                        const VkDeviceCreateInfo* create_info,
+                                        const VkAllocationCallbacks* allocator /*allocator*/,
+                                        VkDevice* device) {
+    auto* layer_create_info = absl::bit_cast<VkLayerDeviceCreateInfo*>(create_info->pNext);
+
+    while (layer_create_info != nullptr &&
+           (layer_create_info->sType != VK_STRUCTURE_TYPE_LOADER_DEVICE_CREATE_INFO ||
+            layer_create_info->function != VK_LAYER_LINK_INFO)) {
+      layer_create_info = absl::bit_cast<VkLayerDeviceCreateInfo*>(layer_create_info->pNext);
+    }
+
+    if (layer_create_info == nullptr) {
+      return VK_ERROR_INITIALIZATION_FAILED;
+    }
+
+    PFN_vkGetInstanceProcAddr next_get_instance_proc_addr_function =
+        layer_create_info->u.pLayerInfo->pfnNextGetInstanceProcAddr;
+    PFN_vkGetDeviceProcAddr next_get_device_proc_addr_function =
+        layer_create_info->u.pLayerInfo->pfnNextGetDeviceProcAddr;
+
+    // Advance linkage for next layer
+    layer_create_info->u.pLayerInfo = layer_create_info->u.pLayerInfo->pNext;
+
+    // Need to call vkCreateInstance down the chain to actually create the
+    // instance, as we need it to be alive in the create instance dispatch table.
+    auto create_device_function = absl::bit_cast<PFN_vkCreateDevice>(
+        next_get_instance_proc_addr_function(VK_NULL_HANDLE, "vkCreateDevice"));
+    VkResult result = create_device_function(physical_device, create_info, allocator, device);
```

Comment:
Similar question here: Should we check "result" before proceeding with the CreateDeviceDispatchTable call?

Reply:
Looks good.